### PR TITLE
bigquery lineage & stats via remote service

### PR DESCRIPTION
**Pull request by vidma**
- [x] support external SQL lineage service
- [x] specs/tests with mocked:
  - [x]  bigquery
  - [x] mocked SQL lineage service - vcr recorder to record request=>response of client
- [x] simple spec to test lineage was generated
- [x] shorten DS names (bug?)

P.S. the most interesting thing for now is (still simplistic) specs. the specs could even parse the offline file (.jsonl) and do more sophisticated checks, but I wanted to keep them simple for now.

the code will need to be further cleaned up.

### CODE REVIEW

File: kensu/google/cloud/bigquery/job/bigquery_stats.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        for s in vs.keys():
+            v = stats[k + "_" + s]
+            if v.__class__ in [datetime.date, datetime.datetime, datetime.time]:
+                v = int(v.strftime("%s") + "000")
+            r[k + "." + s] = v
+    return r
+
+
+def generate_fallback_stats_queries(table):
+    logging.debug('Got empty statistic listing from remote service, proceeding with fallback statistic list')
```
Comment (leszekgruchala) [inline kensu/google/cloud/bigquery/job/bigquery_stats.py:37]: Nitpic: I think this line belongs to the caller of this function.

File: kensu/numpy/__init__.py [hunk 2]
```diff
@@ -417,9 +417,18 @@ def wrapper(*args, **kwargs):
...
-        orig_ds = eventually_report_in_mem(kensu.extractors.extract_data_source(nd, kensu.default_physical_location_ref,
+        try:
+            orig_ds = eventually_report_in_mem(kensu.extractors.extract_data_source(nd, kensu.default_physical_location_ref,
                                                                                 logical_naming=kensu.logical_naming))
-        orig_sc = eventually_report_in_mem(kensu.extractors.extract_schema(orig_ds, nd))
+            orig_sc = eventually_report_in_mem(kensu.extractors.extract_schema(orig_ds, nd))
+        except Exception:  # thrown by default extract_data_source/extract_schema
+            msg = "Kensu numpy.array got an unexpected argument which can not be fully tracked (yet): " + str(type(nd))
+            logging.warning(msg)
+            print(msg)
```
Comment (leszekgruchala) [inline kensu/numpy/__init__.py:427]: print?

File: kensu/google/cloud/bigquery/job/bigquery_stats.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    if input_filters is not None and len(input_filters) > 0:
+        filters = f"WHERE {' AND '.join(input_filters)}"
+    stats_query = f"select {selector}, sum(1) as nrows from `{str(table_ref)}` {filters}"
+    logging.debug(f"stats query for table {table_ref}: {stats_query}")
+    sts = client.query(stats_query)
+    sts_result = sts.result()
+    stats = None
+    for row in sts_result:
+        stats = row  # there is only one anyway
+    r['nrows'] = stats['nrows']
```
[the remark this answers is not in the document]
Comment (vidma) [reply kensu/google/cloud/bigquery/job/bigquery_stats.py:26]: yes.
the other stats below are all `col.statName`  while this one is not column dependent, so need to be extracted diffently.

indeed further cleanup/rename would be useful here...

File: kensu/numpy/__init__.py (hunk 2 above)
Comment (vidma) [reply kensu/numpy/__init__.py:427]: pytest was hiding my logs, finally found a way to fight it `--log-cli-level=INFO` :)

**Review by leszekgruchala — APPROVED** [review]
💪

**Pull request merged**